migrations/lib/importer/steps/base/site_settings.rb

```diff
@@ -0,0 +1,231 @@
+      end
+
+      # TODO: Replace with logging when available
+      @log_messages.each { |message| puts "      #{message}" }
+    end
+
+    private
+
+    # Override in subclasses to skip rows that should be handled in a later step.
+    def skip_row?(_row)
+      false
+    end
+
+    # Override in subclasses to adapt/normalize values by type.
+    def transform_value(_setting_name, value, _type)
+      value
+    end
+
+    private
+
+    def fetch_rows
+      @intermediate_db.query <<~SQL
+        SELECT name, value, last_changed_at, import_mode
+        FROM site_settings
+        ORDER BY ROWID
+      SQL
+    end
+
+    def process_rows(rows)
+      rows.each do |row|
+        @stats.reset
+        before_fail_count = @failed_settings.size
+
+        import_row(row) if row_importable?(row)
+
+        # Only advance the progress bar if the row didn't get deferred due to failure.
+        update_progressbar if before_fail_count == @failed_settings.size
+      end
+
+      # Re-attempt failed updates (often due to dependency ordering).
+      failures_before_retry = @failed_settings.size
+      retry_failed_updates
+      update_progressbar(increment_by: failures_before_retry)
+    end
+
+    def row_importable?(row)
+      row in { name:, import_mode: }
+      name_symbol = row[:name].to_sym
+      setting = @settings_index[name_symbol]
+
+      if setting.nil?
+        log_warning "Ignoring unknown site setting: #{name}"
+        return false
+      end
+
+      if DISALLOWED_SITE_SETTINGS.include?(name_symbol)
+        log_warning "Ignoring disallowed site setting: #{name}"
+        return false
+      end
+
+      if setting[:themeable]
+        log_warning "Can't modify themeable site setting: #{name}"
+        return false
+      end
+
+      if SiteSetting.shadowed_settings.include?(name)
+        log_warning "Can't modify shadowed site setting: #{name}"
+        return false
+      end
+
+      type = setting[:type]
+      if import_mode == Enums::SiteSettingImportMode::APPEND && !type.end_with?("list")
+        log_warning "Can't append to a site setting of type '#{type}': #{name}"
+        return false
+      end
+
+      true
+    end
+
+    def import_row(row)
+      row in { name:, value:, last_changed_at:, import_mode: }
+      setting = @settings_index[name.to_sym]
+
+      case import_mode
+      when Enums::SiteSettingImportMode::AUTO
+        auto_import(setting, value, last_changed_at)
+      when Enums::SiteSettingImportMode::APPEND
+        append_to_list(setting, value)
+      when Enums::SiteSettingImportMode::OVERRIDE
+        set_and_log(setting, transform_value(name, value, setting[:type]))
+      end
+    end
+
+    def auto_import(setting, value, last_changed_at)
+      name = setting[:setting]
+      last_changed_at ||= Time.now
+      existing_last_changed_at = @last_change_dates[name]
+
+      if !existing_last_changed_at || existing_last_changed_at < last_changed_at
```

Comment:
The `auto` import mode updates a setting only if the existing setting value is older than the value we want to import. We could make this a bit more clever in the future, but that's it for now.